File: core/data/src/commonMain/kotlin/com/shelldocs/core/data/markdown/MarkdownParser.kt
package com.shelldocs.core.data.markdown

import com.shelldocs.core.domain.entity.document.*

/**
 * Kotlin port of the original backend `markdown-parser.ts`: converts raw
 * Markdown into semantic blocks (heading, paragraph, list, code, quote),
 * plus plain text for search and a content hash for deduplication.
 */
class MarkdownParser {

    fun parse(rawMarkdown: String): ParsedMarkdown {
        val blocks = parseBlocks(rawMarkdown)
        val plainText = blocks.joinToString("\n") { blockText(it) }.trim()
        return ParsedMarkdown(
            content = DocumentContent(schemaVersion = SCHEMA_VERSION, blocks = blocks),
            plainText = plainText,
            contentHash = ContentHasher.hash(rawMarkdown),
        )
    }

    private fun parseBlocks(rawMarkdown: String): List<ContentBlock> {
        val blocks = mutableListOf<ContentBlock>()
        val lines = rawMarkdown.lines()
        var index = 0

        while (index < lines.size) {
            val line = lines[index]
            val trimmed = line.trim()
            when {
                trimmed.isEmpty() -> index++

                isTableRow(trimmed) && index + 1 < lines.size && TABLE_SEPARATOR.matches(lines[index + 1].trim()) -> {
                    val headers = splitTableRow(trimmed)
                    index += 2 // header + separator
                    val rows = mutableListOf<List<String>>()
                    while (index < lines.size && isTableRow(lines[index].trim())) {
                        rows += splitTableRow(lines[index].trim())
                        index++
                    }
                    blocks += TableBlock(headers = headers, rows = rows)
                }

                trimmed.startsWith("```") -> {
                    val language = trimmed.removePrefix("```").trim()
                    val code = mutableListOf<String>()
                    index++
                    while (index < lines.size && !lines[index].trim().startsWith("```")) {
                        code += lines[index]
                        index++
                    }
                    index++ // closing fence
                    blocks += CodeBlock(language = language, code = code.joinToString("\n"))
                }

                HEADING.matches(trimmed) -> {
                    val match = HEADING.find(trimmed)!!
                    blocks += HeadingBlock(
                        level = match.groupValues[1].length,
                        text = match.groupValues[2].trim(),
                    )
                    index++
                }

                trimmed.startsWith("> ") || trimmed == ">" -> {
                    val quote = mutableListOf<String>()
                    while (index < lines.size && lines[index].trim().let { it.startsWith(">") }) {
                        quote += lines[index].trim().removePrefix(">").trim()
                        index++
                    }
                    blocks += QuoteBlock(text = quote.joinToString(" ").trim())
                }

                BULLET_ITEM.matches(trimmed) -> {
                    val items = mutableListOf<String>()
                    while (index < lines.size && BULLET_ITEM.matches(lines[index].trim())) {
                        items += BULLET_ITEM.find(lines[index].trim())!!.groupValues[1].trim()
                        index++
                    }
                    blocks += ListBlock(style = ListStyle.BULLET, items = items)
                }

                ORDERED_ITEM.matches(trimmed) -> {
                    val items = mutableListOf<String>()
                    while (index < lines.size && ORDERED_ITEM.matches(lines[index].trim())) {
                        items += ORDERED_ITEM.find(lines[index].trim())!!.groupValues[1].trim()
                        index++
                    }
                    blocks += ListBlock(style = ListStyle.ORDERED, items = items)
                }

                else -> {
                    val paragraph = mutableListOf<String>()
                    while (index < lines.size && lines[index].trim().isNotEmpty() &&
                        !isBlockStart(lines[index].trim())
                    ) {
                        paragraph += lines[index].trim()
                        index++
                    }
                    blocks += ParagraphBlock(text = paragraph.joinToString(" "))
                }
            }
        }
        return blocks
    }

    private fun isBlockStart(trimmed: String): Boolean =
        trimmed.startsWith("```") || trimmed.startsWith(">") || isTableRow(trimmed) ||
            HEADING.matches(trimmed) || BULLET_ITEM.matches(trimmed) || ORDERED_ITEM.matches(trimmed)

    private fun isTableRow(trimmed: String): Boolean =
        trimmed.startsWith("|") && trimmed.endsWith("|") && trimmed.length > 1

    private fun splitTableRow(trimmed: String): List<String> =
        trimmed.removePrefix("|").removeSuffix("|").split("|").map { it.trim() }

    private fun blockText(block: ContentBlock): String = when (block) {
        is HeadingBlock -> block.text
        is ParagraphBlock -> stripInline(block.text)
        is ListBlock -> block.items.joinToString("\n") { stripInline(it) }
        is CodeBlock -> block.code
        is QuoteBlock -> stripInline(block.text)
        is TableBlock -> (listOf(block.headers) + block.rows).joinToString("\n") { it.joinToString(" ") }
    }

    private fun stripInline(text: String): String =
        text.replace(INLINE_MARKERS, "")

    private companion object {
        const val SCHEMA_VERSION = 1
        val HEADING = Regex("^(#{1,6})\\s+(.*)$")
        val BULLET_ITEM = Regex("^[-*+]\\s+(.*)$")
        val ORDERED_ITEM = Regex("^\\d+[.)]\\s+(.*)$")
        val TABLE_SEPARATOR = Regex("^\\|?\\s*:?-+:?\\s*(\\|\\s*:?-+:?\\s*)*\\|?$")
        val INLINE_MARKERS = Regex("[*_`]")
    }
}

File: backend/src/markdown-parser.ts
import {createHash} from "node:crypto";

export type DocumentBlock =
    | { type: "heading"; level: number; text: string }
    | { type: "paragraph"; text: string }
    | { type: "list"; style: "unordered" | "ordered"; items: string[] }
    | { type: "code"; language?: string; code: string }
    | { type: "blockquote"; text: string };

export type ParsedDocument = {
    title?: string;
    contentJson: {
        type: "document";
        schemaVersion: 1;
        blocks: DocumentBlock[];
    };
    plainText: string;
    tableOfContents: Array<{ level: number; text: string }>;
    contentHash: string;
};

export function parseMarkdown(markdown: string): ParsedDocument {
    const lines = markdown.replace(/\r\n/g, "\n").split("\n");
    const blocks: DocumentBlock[] = [];
    const tableOfContents: Array<{ level: number; text: string }> = [];
    const plainTextParts: string[] = [];
    let title: string | undefined;
    let index = 0;

    while (index < lines.length) {
        const line = lines[index];
        const trimmed = line.trim();

        if (!trimmed) {
            index += 1;
            continue;
        }

        if (trimmed.startsWith("```")) {
            const language = trimmed.slice(3).trim() || undefined;
            const codeLines: string[] = [];
            index += 1;
            while (index < lines.length && !lines[index].trim().startsWith("```")) {
                codeLines.push(lines[index]);
                index += 1;
            }
            blocks.push({type: "code", language, code: codeLines.join("\n")});
            plainTextParts.push(codeLines.join("\n"));
            index += 1;
            continue;
        }

        const heading = /^(#{1,6})\s+(.+)$/.exec(trimmed);
        if (heading) {
            const level = heading[1].length;
            const text = heading[2].trim();
            blocks.push({type: "heading", level, text});
            tableOfContents.push({level, text});
            plainTextParts.push(text);
            title ??= level === 1 ? text : undefined;
            index += 1;
            continue;
        }

        if (/^>\s?/.test(trimmed)) {
            const text = trimmed.replace(/^>\s?/, "");
            blocks.push({type: "blockquote", text});
            plainTextParts.push(text);
            index += 1;
            continue;
        }

        if (/^[-*+]\s+/.test(trimmed)) {
            const items: string[] = [];
            while (index < lines.length && /^[-*+]\s+/.test(lines[index].trim())) {
                items.push(lines[index].trim().replace(/^[-*+]\s+/, ""));
                index += 1;
            }
            blocks.push({type: "list", style: "unordered", items});
            plainTextParts.push(...items);
            continue;
        }

        if (/^\d+\.\s+/.test(trimmed)) {
            const items: string[] = [];
            while (index < lines.length && /^\d+\.\s+/.test(lines[index].trim())) {
                items.push(lines[index].trim().replace(/^\d+\.\s+/, ""));
                index += 1;
            }
            blocks.push({type: "list", style: "ordered", items});
            plainTextParts.push(...items);
            continue;
        }

        const paragraphLines = [trimmed];
        index += 1;
        while (
            index < lines.length &&
            lines[index].trim() &&
            !/^(#{1,6})\s+/.test(lines[index].trim()) &&
            !/^[-*+]\s+/.test(lines[index].trim()) &&
            !/^\d+\.\s+/.test(lines[index].trim()) &&
            !/^>\s?/.test(lines[index].trim()) &&
            !lines[index].trim().startsWith("```")
            ) {
            paragraphLines.push(lines[index].trim());
            index += 1;
        }
        const text = paragraphLines.join(" ");
        blocks.push({type: "paragraph", text});
        plainTextParts.push(text);
    }

    const plainText = plainTextParts.join("\n").trim();
    return {
        title,
        contentJson: {
            type: "document",
            schemaVersion: 1,
            blocks,
        },
        plainText,
        tableOfContents,
        contentHash: `sha256:${createHash("sha256").update(markdown).digest("hex")}`,
    };
}
